core/store/src/db.rs

```diff
@@ -89,14 +167,16 @@ pub trait Database: Sync + Send {
     ///
     /// You most likely will want to use [`refcount::get_with_rc_logic`] to
     /// properly handle reference-counted columns.
-    fn get_raw_bytes(&self, col: DBCol, key: &[u8]) -> io::Result<Option<Vec<u8>>>;
+    fn get_raw_bytes(&self, col: DBCol, key: &[u8]) -> io::Result<Option<DBBytes<'_>>>;
 
     /// Returns value for given `key` forcing a reference count decoding.
     ///
     /// **Panics** if the column is not reference counted.
-    fn get_with_rc_stripped(&self, col: DBCol, key: &[u8]) -> io::Result<Option<Vec<u8>>> {
+    fn get_with_rc_stripped(&self, col: DBCol, key: &[u8]) -> io::Result<Option<DBBytes<'_>>> {
         assert!(col.is_rc());
-        Ok(self.get_raw_bytes(col, key)?.and_then(crate::db::refcount::strip_refcount))
+        Ok(self
+            .get_raw_bytes(col, key)?
+            .and_then(|bytes| refcount::decode_value_with_rc(&bytes).0.map(DBBytes::from)))
```

Comment:
We still force an alloc here, it seems it can be avoided?

https://github.com/facebook/rocksdb/blob/9e583711144f580390ce21a49a8ceacca338fcd5/include/rocksdb/slice.h#L166

It feels relatively important that, if we go with 0-copy API, that rc is also 0-copy

Reply:
Sadly that method is available neither in Rust RocksDB crate nor in librocksdb C API.  The only thing the API allows us to do is get the slice but changing it is not possible.

Reply:
Hm, my gut feel would be to not proceed here: rc feels like the most important case for us, and if we can't avoid memcpy there, probably it doesn't really matter? Like, we gain some perf (which I expect to be <1%), but at the cost of bifurcated architecture (we either do, or not do copy, for no good reason).

Alternatively, the `_with(&self, cb: &dyn Fn(&[u8]))` API should work here. A virtual call probably won't be problematic in the grand scheme of things, given that we already go through a C API which I think allocates the PinnedSlice iteslf on the heap?


:thinking: I am also not sure if it's a good long-term plan to expose PinnableSlice-shaped API. Like, we entertain the thought of writing our own storage, and it's not clear to me whether that would have pin API (like, if we are doing flat hash-map on disk where we just mmap stuff, we probably don't want to have caches on top of mmap, and pining the whole mmaped region feels suboptimal).

---

Looking more holistically, I wonder if intefacing to rocks via C API is a good idea at all, and if, in an ideal world with infinite time, we'd just https://github.com/dtolnay/cxx the C++ API?

Reply:
We do `get_ser` on quite a few non-reference-counted columns so I don’t think it doesn’t really matter.

This also reduces number of places where copying happens thus making thinking of a solution potentially easier.  The most obvious one is someone getting rocksdb_pinnableslice_remove_suffix function added to the C API.  Long term it may also be that we don’t even have get_with_rc_stripped and instead rely on strongly typed objects where removal of refcount happens outside of the get method.

Even without all that this can still be solved.  I’ll have a fix in a bit though you won’t like it.

> I am also not sure if it's a good long-term plan to expose PinnableSlice-shaped API.

I would argue with the statement that we expose such API.  Yes, DBBytes does have that API but it’s purpose is a translation layer between the database representation and what clients of the Database traits want to actually use.  Other Database implementation have an option to return a vector just like we used to.

> Looking more holistically, I wonder if intefacing to rocks via C API is a good idea at all, and if, in an ideal world with infinite time, we'd just https://github.com/dtolnay/cxx the C++ API?

Yes, of course.  But I’m not gonna do it. ;)

Reply:
:thinking: yeah, I think I am convinced that this is the shape of the API we want here. Especially in the `CompiledContractCache` case, it does look like we want `DbBytes` long-term there, as we want to zero-copy deserialize that.

And, well, jakmeier benchmarks show that this looks like an improvement, and we did manage to avoid copies in rc columns.